### PR TITLE
MNCH 518 whatsapp message buttons

### PR DESCRIPTION
**Pull request by rudigiesler**
## Purpose
We want to be able to define a list of up to 3 buttons for whatsapp messages, to be used for interactive messages

## Approach
We create a new StreamBlock for whatsapp messages, named "buttons". Here we can add up to 3 buttons of varying types.

Currently we have "next message" and "go to page" types, but we can always add more if the need arises.

We also add a data migration, to change all the `next_prompt`s to buttons, and then we removed the `next_prompt` field

image

### CODE REVIEW

File: home/tests/test_models.py
Comment (jerith) [inline home/tests/test_models.py:158]: Do we need tests for the new `buttons` field in here?

**Review by jerith — COMMENTED** [review]
I don't see any changes to the import/export code here. Does that mean it'll still try to import `next_prompt`?

File: CHANGELOG.md
Comment (jerith) [reply CHANGELOG.md:15]: I've just checked. The exporter handles `next_prompt`, but the (current) importer apparently ignores it. Merging this branch into the import/export refactor PR gives me test failures for:
* `next_prompt` not being preserved across import->export with new importer (kind of expected because it's been removed from the model, but I'd expect the export to fail because of the missing field -- apparently we explicitly ignore those errors, which we probably shouldn't)
* (empty) `buttons` field not being present in export->import with current importer (expected because the current importer builds the content blocks generically instead of using the appropriate block models)

File: pyproject.toml
Comment (jerith) [inline pyproject.toml:7]: Do we really need to bump the major version for this?

Comment (rudigiesler) [reply pyproject.toml:7]: Any existing services using the API will be looking at `next_prompt` in the API, they need to change to look in the `buttons` section instead, in a completely different format (there may be multiple buttons, they might not all be next buttons). So I would say for the API users, it is a major change.

The other thing we could do is version the API. So do something like the existing API returns the first next message button as `next_prompt`, and then we have a new version that returns the full `buttons` detail. I'm not sure if that's worth the effort though.

File: CHANGELOG.md
Comment (rudigiesler) [reply CHANGELOG.md:15]: Yeah so import/export will need to be changed, but that's a bit difficult while it's still being worked on. I don't really want to make the changes to the old import/export, if it's going to be replaced anyway. Maybe we could hold off on this PR until that work is complete, then rebase/merge, and add to the new importer?

File: home/tests/test_models.py
Comment (rudigiesler) [reply home/tests/test_models.py:158]: I've added tests for the validations on the individual button types. I'm unsure how to add tests here for compound values, adding to the dictionary doesn't work. The only test I would want to add here is to ensure you can't add more than 3 buttons

Comment (rudigiesler) [reply home/tests/test_models.py:158]: Okay I've figured out the compound values, added a test for max 3 buttons.

File: home/import_content_pages.py
Comment (rudigiesler) [inline home/import_content_pages.py:162]: This logic was being duplicated in `add_message_to_shadow_content_page_from_row`, so I've removed it from here and used that instead.

File: CHANGELOG.md
Comment (jerith) [inline CHANGELOG.md:14]: If we're deprecating `next_prompt` instead of removing it, do we still want this migration?

File: home/content_import_export.py
Comment (jerith) [inline home/content_import_export.py:785]: Out of scope for this PR, but should we perhaps copy and refactor the exporter like we did for the importer?

File: home/import_content_pages.py
Comment (jerith) [inline home/import_content_pages.py:83]: Where did this assumption break? It's never actually been valid, but having the import fail is probably better than silently reparenting pages.

File: home/tests/test_content_import_export.py
Comment (jerith) [inline home/tests/test_content_import_export.py:332]: Would it make sense to combine this, `normalise_related_page_ids`, and `normalise_body_ids` into a more generic `normalise_block_ids` filter? (I'm happy to give that a try, probably in a separate PR.)

File: home/tests/test_content_import_export.py
Comment (jerith) [inline home/tests/test_content_import_export.py:741]: Do we still have tests for `next_prompt` now that it's not gone completely? Do we support mixing buttons and `next_prompt` in the same message? (If not, we should probably enforce that somehow.)

**Review by jerith — COMMENTED** [review]
I think we're going to need explicitly versioned import (and maybe export?) formats. Doing that in the exported files will be a bit clunky, but we can probably determine the version based on the set of headers/keys we get. Any thoughts?

File: CHANGELOG.md
Comment (rudigiesler) [reply CHANGELOG.md:14]: Yes I think so, it will allow people to easily move over to using the new format, instead of content admins having to manually go through all the content and add buttons.

File: home/import_content_pages.py
Comment (rudigiesler) [reply home/import_content_pages.py:83]: This broke when using data on my local system, which is from an export of a production system. I think it is better to fail here, so maybe I'll change it back, although that will break backwards compatibility for old exports.

File: home/tests/test_content_import_export.py
Comment (rudigiesler) [reply home/tests/test_content_import_export.py:741]: Hmm that's true. We do support mixing them (they should be treated as just separate things, and the application has to decide which one to use, to allow a smooth changeover), so I'll update this test to check that all combinations can work

File: home/tests/test_content_import_export.py
Comment (rudigiesler) [reply home/tests/test_content_import_export.py:332]: Yeah happy for you to give it a try in a separate PR. I don't think it's high priority though, since we've only got 2 instances of us needing to do this (I prefer to wait until we have 3), and it's in tests not production code, but I think blocks are well enough defined in wagtail that we can make something generic.

File: CHANGELOG.md
Comment (jerith) [reply CHANGELOG.md:14]: So existing `next_prompt`s will be migrated, but then any new ones (added or imported) will need to be manually changed over at some point before `next_prompt` goes away? Is it worth adding a management command or something to do that?

**Review by jerith — APPROVED** [review]

File: CHANGELOG.md
Comment (rudigiesler) [reply CHANGELOG.md:14]: Migrating at this point, when we know `buttons` is empty, is easy to do. Doing it later on is much more difficult, eg. what do you do if there are already 3 buttons, and will require human intervention, so I'm happy leaving that for now, and working through any specific use cases (and maybe creating a management command at that point once we know more)

**Pull request merged**
